### ISSUE TITLE
Sparse vector dense vector product is painfully slow

### ISSUE THREAD

**Issue by usamec** [vbarrielle/sprs#125]
Given:
```
let w = Array::range(0., 10., 0.00001);
let x = CsVec::new(1000000, vec![0, 200000, 800000], vec![1., 2., 3.]);
```

It is much faster to do:
```
x.iter().map(|(index, num)| num * w[index]).sum();
```
than
```
x.dot(&w);
```

This should either be fixed, or at least documented somewhere.

**Comment by usamec** [comment]
Same case also happens when multiplying sparse matrix by vector:
```
        let w = Array::range(0., 10., 0.00001);
        let a = CsMat::new((3, 1000000),
                               vec![0, 2, 4, 5],
                               vec![0, 1, 0, 2, 2],
                               vec![1., 2., 3., 4., 5.]);
        let r = a.mul(&w);
```

**Comment by vbarrielle** [comment]
On Wednesday 20 December 2017 23:24:22 usamec wrote:
> Same case also happens when multiplying sparse matrix by vector:
> ```
>         let w = Array::range(0., 10., 0.00001);
>         let a = CsMat::new((3, 1000000),
>                                vec![0, 2, 4, 5],
>                                vec![0, 1, 0, 2, 2],
>                                vec![1., 2., 3., 4., 5.]);
>         let r = a.mul(&w);
> ```

Thanks for the report, I'll have a look at why it's slower.

**Comment by vbarrielle** [comment]
So for the sparse vector times dense vector case, the problem is that the `dot` method assumes both inputs are sparse, and thus does not make use of the random indexing ability of dense vectors. It should be possible to dynamically switch to a specialized implementation for a dense rhs.

**Comment by vbarrielle** [comment]
I've looked into the sparse matrix/dense vector multiplication issue, there are two separate problems:

- the csr-dense by row product subroutine I had implemented was trying to do some blocking on the product as can de done in dense-dense matrix multiplication but it looks like that was a bad idea. A more classic implementation performs decently (went from 8,469,111 ns on your example down to 84 ns).
- for very thin matrices like the one in your example, it is much more performant to compute the product by iterating on the rhs by column (on your example this gets down to 26 ns). I'll have to determine a thinness factor to switch on the more performant method.

The current status is that I have a [branch](https://github.com/vbarrielle/sprs/tree/faster_sparse_dense_dot_prod) with a `dot_dense` function and a correction to the csr/dense product. What remains to be done is that the `dot` function should automatically call the `dot_dense` function when needed, and I should write a better subroutine selector for sparse/dense products.
